src/gsi/setupq.f90

```diff
@@ -549,10 +579,36 @@ subroutine setupq(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
      rmaxerr=max(small2,rmaxerr)
      errorx =(data(ier,i)+dprpx)*qsges
 
-! Interpolate guess moisture to observation location and time
-     call tintrp31(ges_q,qges,dlat,dlon,dpres,dtime, &
-        hrdifsig,mype,nfldsig)
-    
+! qges: Interpolate guess moisture to observation location and time 
+
+     if (.not. ( hofx_2m_sfcfile .and. landsfctype) ) then 
+        call tintrp31(ges_q,qges,dlat,dlon,dpres,dtime, &
+           hrdifsig,mype,nfldsig)
+     else
+        ! only use land locations
+        if (int(data(idomsfc,i)) .NE. 1  ) muse(i) = .false.
+
+        call tintrp2a11(ges_q2m,qges,dlat,dlon,dtime,hrdifsig,mype,nfldsig)
+
+        ! terrain correction: assume RH_zo = RH_zm, and correct T with 
+        ! same lapse rate as used for T2m terrain correction
+
+        delta_z = data(istnelv,i) - data(izz,i) ! obs -model
+
+        do jj=1,nfldsig
+           ! qsat in model at height of obs
+           call genqsat(qg2m_o(:,:,jj),ges_t2m(:,:,jj)+delta_z*T_lapse,ges_ps(:,:,jj),lat2,lon2,1,ice,iderivative)
+        enddo
+
+        call tintrp2a11(qg2m_o,qsges_o,dlat,dlon,dtime,hrdifsig,mype,nfldsig)
+        q_delta_terrain = (qsges/qsges_o - 1)*qob
+        qob = qob * ( qsges/qsges_o)
+
+        !update the station elevation
+        data(istnelv,i) = data(izz,i)
+
+     endif
+
```

Comment:
Please discard my previous comment on setupq.f90. When viewed from the surface/soil perspective, I got what the code snippet 588-608 does. 
One question on the calculation of qsges_o:  
Currently, you use the height difference at the obs location and apply its terrain adjustment to the full 2D get_t2m, compute corresponding 2D qq2m_o, interpolate a final qsges_o. 
An alternative way is to interpolate a t2mges, apply terrain adjustment to get t2mges_o, then use this t2mges_o to get a final qsges_o.
Do you think the two methods are equivalent?

Reply:
I think the method you lay out here is a probably better, and they'll only give the same answer if the obs are in the center of the grid cell. However, I did it this way as the genqsat routine is expecting input/output of size (lat2,lon2,nsig), so I'd have to rewrite it to make it flexible enough to accept a scalar t2mges_o. 

FYI: The UKMO terrain-corrects their q obs by assigning a fixed lapse rate to q (like I implemented for T). At some point, I'll likely look into whether we can simplify this - to date, I've only looked at one analysis time and the method I implemented was quite effective at remove a bias that was correlated with terrain mismatch for q (could have been a coincidence though!). 

I think I missed your earlier comment on setupq, and can't find it now. Please let me know if I've missed anything else.

Reply:
It is interesting to know that UKMO uses a fixed lapse rate for q.  Do they have any publications on this? 
I like your way of dealing with the terrain effect for q.

I compared the two ways of computing qsges_o and would think they are not equivalent as the equation of qs(T, P) is not linear from T to qs. It may be better to use the second method. 

we may compute scalar qs (pseudo codes) as follows:

```
allocate qsges_o(1,1,1), ges_t2m_o(1,1,1), ges_ps_o(1,1,1)
genqsat(qsges_o,ges_t2m_o, ges_ps_o, 1,1,1,ice,derivative)
```

I think it will work but I did not try it.

Reply:
Thanks for the reference!